usr/src/uts/common/fs/zfs/metaslab.c

```diff
@@ -2854,9 +2933,11 @@ metaslab_group_alloc_normal(metaslab_group_t *mg, zio_alloc_list_t *zal,
 		 * We have just failed an allocation attempt, check
 		 * that metaslab_should_allocate() agrees. Otherwise,
 		 * we may end up in an infinite loop retrying the same
-		 * metaslab.
+		 * metaslab. Keep in mind, it might have happened due
+		 * to an ongoing trim.
```

Comment:
Ah, yes, you are correct. How about I add a weight modification into the metaslab selector that applies a penalty based based on the amount of space in it that needs to be trimmed. Thus metaslabs that need to trim a lot will appear a lot less appealing to allocate from, but if push comes to shove, we could do it. It also keeps the impact on the normal weighting in check by making occasional small trims apply a comparatively tiny penalty.